### PR TITLE
Npm module

### PR DESCRIPTION
**Pull request by scottnath**
@sergesemashko @ericduran 

Could you two look this over and give me feed back please?

The tests which are not skipped are all working.

thanks,
Scott

### CODE REVIEW

**Comment by scottnath** [comment]
@sergesemashko this is the pull request

**Comment by sergesemashko** [comment]
thanks

File: index.js [hunk 1]
```diff
@@ -1,65 +1,27 @@
...
-    absolute: file.path,
-    relative: file.path.replace(process.cwd() + '/', ''),
-    folder: file.path.replace(process.cwd() + '/', '').split('/').slice(0, -1).join('/'),
-    inner: file.path.replace(process.cwd() + '/', '').split('/').slice(1, -1).join('/')
-  };
-  return paths;
-}
+ * @param {Object} options
+*/
+var patternImporter = function patternImporter (options) {
```
Comment (sergesemashko) [inline index.js]: Variable definition is excess here, it can be just: 

``` js
function patternImporter (options) {...}
```

File: index.js [hunk 2]
```diff
@@ -71,173 +33,15 @@ module.exports = function (options) {
...
-        configYml.includes = meta.includes
-      }
-
-      // source will be the location of the pattern in this project
-      configYml.source = paths.folder;
-
-      // source of the DATA used - NATH: this section will change with addition of func to add LIVE/LOCAL data
-      configYml.data = 'source';
+    // convert single pattern
+    importSinglePattern.importSinglePattern(file, options, compiledPatterns);
```
Comment (sergesemashko) [inline index.js]: `importSinglePattern.importSinglePattern` - I would suggest to have different names for object and method name.

File: lib/css-utils.js [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+var utils = require('../lib/utils'),
+    sasser = require('./css-compilers/sass-compiler');
+
+/**
+ * Returns an object with the string-name of the style file to compile and our desired css-compiling engine
+ * @function determineCssCompiler
+ *
+ */
+var determineCssCompiler = function determineCssCompiler (options, patternObject) {
```
Comment (sergesemashko) [inline lib/css-utils.js]: Variable definition can be simplified to `function determineCssCompiler (options, patternObject) {`

File: lib/css-utils.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+  return cssCompilerData;
+}
+
+/**
+ * Sends out style file to the correct compiler
+ * @function compileCss
+ *
+ */
+var compileCss = function compileCss (paths, cssCompilerData) {
```
Comment (sergesemashko) [inline lib/css-utils.js]: same as above

File: lib/css-utils.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  return cssOutput;
+
+}
+
+/**
+ * Creates the file name for our css file
+ * @function cssFileName
+ *
+ */
+var cssFileName = function cssFileName (name) {
```
Comment (sergesemashko) [inline lib/css-utils.js]: same as above

File: lib/css-utils.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+  return String(name) + '.css';
+
+}
+
+/**
+ * Writes compiled css to a file
+ * @function writeCss
+ *
+ */
+var writeCss = function writeCss (dest, cssCompiledContents) {
```
Comment (sergesemashko) [inline lib/css-utils.js]: same as above

File: lib/css-utils.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+ */
+var compileCss = function compileCss (paths, cssCompilerData) {
+
+  if(!cssCompilerData.compilingEngine){
+    throw new gutil.PluginError(PLUGIN_NAME, 'Somehow a compilingEngine is not listed in the cssCompilerData object. Probably your fault.');
+  }
+
+  var cssOutput;
+
+  if (cssCompilerData.compilingEngine === 'sass') {
```
Comment (sergesemashko) [inline lib/css-utils.js]: These if's statements can be replaced by switch:

``` js
switch(cssCompilerData.compilingEngine) {
  case "sass":
    ...
  break;
  ...
  default: ...
}

```

File: lib/import-single-pattern.js [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+ * @returns {String} patternFiles.filesToCopy[file].src   relative path to file source
+ * @returns {String} patternFiles.filesToCopy[file].dest   relative path to file destination
+ * @returns {Object} patternFiles.includedFiles   Contains all css & js files required for this pattern (including included-pattern files)
+ * @returns {Array} patternFiles.includedFiles.css  relative path to css file's destination
+ * @returns {Array} patternFiles.includedFiles.js  relative path to js file's destination
+ * @returns {String} patternFiles.patternCss  relative path for this pattern's main COMPILED css file
+ * @returns {String} patternFiles.patternScript  relative path for this pattern's main js file
+
+*/
+var getPattern = function getPattern (paths, options, compiledPatterns) {
```
Comment (sergesemashko) [inline lib/import-single-pattern.js]: same as above

File: lib/import-single-pattern.js [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+    };
+
+    // cycle through each of our includes
+    patternObject.includes.forEach(function (include){
+      // create a vinyl version of the include's pattern.yml
+      var includeFile = utils.createFile(path.join(paths.folder, include, options.dataFileName), process.cwd(), path.join(paths.folder, include));
+
+      // compile the pattern, return it's included files
+      var includePatternFiles = getPattern(utils.getFilePaths(includeFile), options, compiledPatterns);
+      if(includePatternFiles.patternCss !== undefined){
```
Comment (sergesemashko) [inline lib/import-single-pattern.js]: should file be checked for existence here?

File: lib/import-single-pattern.js [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+  return patternFiles;
+
+}
+
+/**
+ * Writes pattern files to the file system
+ * @function writePattern
+ *
+ */
+var writePattern = function writePattern (patternFiles) {
```
Comment (sergesemashko) [inline lib/import-single-pattern.js]: same as above

File: lib/pattern-compiler.js [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+ * @param {String} getPatternDestPath
+ * @param {Object} paths
+ *
+ * @returns {Object} patternCompilerData - Info about the file being converted into the pattern's html template
+ * @returns {Object} patternCompilerData.data - The data for the template's variables
+ * @returns {String} patternCompilerData.dest - Relative url for the final HTML template
+ * @returns {String} patternCompilerData.src - Relative url for the template-to-be-converted
+ * @returns {String} patternCompilerData.templateEngine - The name of the templating engine this conversin will use.
+ */
+var determineCompiler = function determineCompiler (options, patternObject, patternDestPath, paths) {
```
Comment (sergesemashko) [inline lib/pattern-compiler.js]: same as above

File: lib/pattern-compiler.js [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+ * note: this can be expanded to select a compiling engine according to whatever templates are available
+ * @param {Object} patternCompilerData - Info about the file being converted into the pattern's html template
+ * @param {Object} patternCompilerData.data - The data for the template's variables
+ * @param {String} patternCompilerData.dest - Relative url for the final HTML template
+ * @param {String} patternCompilerData.src - Relative url for the template-to-be-converted
+ * @param {String} patternCompilerData.templateEngine - The name of the templating engine this conversin will use.
+ *
+ * @returns {String} raw html
+ */
+var compileTemplate = function compileTemplate (patternCompilerData) {
```
Comment (sergesemashko) [inline lib/pattern-compiler.js]: same as above

File: lib/utils.js [hunk 13]
```diff
@@ -0,0 +1,355 @@
...
+    templateDonut: {
+      'twig': './node_modules/pattern-importer/templates/donut.twig'
+    },
+    appFolder: 'app/'
+  }
+  var options = options || {};
+
+  // determine data source
+  options.dataSource = options.dataSource || optionsDefaults.dataSource;
+  if (typeof options.dataSource !== 'string') {
```
Comment (sergesemashko) [inline lib/utils.js:34]: :+1: for validating options

File: lib/utils.js [hunk 14]
```diff
@@ -0,0 +1,355 @@
...
+/**
+ * Returns an array of css files used by the pattern and its included patterns
+ * @function getPatternCssFilesArray
+ *
+ */
+var getPatternCssFilesArray = function getPatternCssFilesArray (patternFiles) {
+  var cssArray = [];
+  if(patternFiles.includedFiles !== undefined){
+    if((patternFiles.includedFiles.css !== undefined) && Array.isArray(patternFiles.includedFiles.css)){
+      patternFiles.includedFiles.css.forEach(function (file) {
```
Comment (sergesemashko) [inline lib/utils.js]: this foreach can be replaced by `cssArray = cssArray.concat(patternFiles.includedFiles.css);`

File: lib/utils.js [hunk 15]
```diff
@@ -0,0 +1,355 @@
...
+
+/**
+ * Returns an array of css files used by the pattern and its included patterns
+ * @function getPatternCssFilesArray
+ *
+ */
+var getPatternCssFilesArray = function getPatternCssFilesArray (patternFiles) {
+  var cssArray = [];
+  if(patternFiles.includedFiles !== undefined){
+    if((patternFiles.includedFiles.css !== undefined) && Array.isArray(patternFiles.includedFiles.css)){
```
Comment (sergesemashko) [inline lib/utils.js]: seems like these 2 if statements can be combined:

``` js
if(patternFiles.includedFiles !== undefined){
    if((patternFiles.includedFiles.css !== undefined) && Array.isArray(patternFiles.includedFiles.css)){
```

also, `(patternFiles.includedFiles.css !== undefined)` is unnecessary as `Array.isArray(patternFiles.includedFiles.css)` is here

File: lib/import-single-pattern.js (hunk 9 above)
Comment (scottnath) [inline lib/import-single-pattern.js]: no, because it doesn't exist yet. All files are written at the end.

**Comment by scottnath** [comment]
@sergesemashko I've fixed all issues. Any other changes before merging?

**Comment by sergesemashko** [comment]
looks good so far, just minor code style comments. 
One more code style thing:
Such kind of constructions

``` js
if (patternFiles.filesToCopy !== undefined)
```

can be replaced by `if (patternFiles.filesToCopy)`

**Comment by scottnath** [comment]
@sergesemashko removed those `!== undefined`

**Comment by sergesemashko** [comment]
looks good for me, @scottnath seems like you have merge conflicts

**Pull request merged**